### PR TITLE
Fix wstest 2.1 , 2.2, 2.3, 2.4, 2.6 : send pong frame with close frame after ping + check is control frame and its longer

### PR DESCRIPTION
**Pull request by folliked**
#1 related

### CODE REVIEW

File: src/Rfc6455/FrameFactory.php [hunk 1]
```diff
@@ -42,6 +42,20 @@ public function createCloseFrame(int $status = Frame::CLOSE_NORMAL, string $reas
         return $frame;
     }
 
+    /**
+     * The payload of Pong must be empty
+     * @return Frame
+     */
+    public function createPongFrame() : Frame
```
Comment (Nek-) [inline src/Rfc6455/FrameFactory.php]: This should take a payload as parameter (the pong answer the ping payload).

File: src/Rfc6455/FrameFactory.php [hunk 2]
```diff
@@ -42,6 +42,20 @@ public function createCloseFrame(int $status = Frame::CLOSE_NORMAL, string $reas
...
+    {
+        $frame = $this->createNewFrame();
+
+        $frame->setOpcode(Frame::OP_PONG);
+        $frame->setPayload('');
```
Comment (Nek-) [inline src/Rfc6455/FrameFactory.php]: As consequence, this should obviously not be empty.

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function supports(Message $message)
+    {
+        return $message->getFirstFrame()->getOpcode() === Frame::OP_PING;
+    }
+
+    public function process(Message $message, MessageProcessor $messageProcessor, ConnectionInterface $socket)
+    {
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createPongFrame(), $socket);
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createCloseFrame(), $socket);
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: A pong does not imply a close !

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        $socket->end();
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: (and so the end is not needed)

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+class PingFrameHandler implements Rfc6455MessageHandlerInterface
+{
+    public function supports(Message $message)
+    {
+        return $message->getFirstFrame()->getOpcode() === Frame::OP_PING;
+    }
+
+    public function process(Message $message, MessageProcessor $messageProcessor, ConnectionInterface $socket)
+    {
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createPongFrame(), $socket);
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: The payload of the ping should be take as parameter (message content).

File: tests/Woketo/Rfc6455/MessageHandler/PingFrameHandlerTest.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+{
+    private $pingMessage;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        // Normal ping frame without payload
+        $this->pingMessage = new Message();
+        $this->pingMessage->addFrame(new Frame(BitManipulation::hexArrayToString(['89', '00'])));
```
Comment (Nek-) [inline tests/Woketo/Rfc6455/MessageHandler/PingFrameHandlerTest.php]: The ping message is related to a specific case (the full message actually). So you should not set it in the setup part !

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function supports(Message $message)
+    {
+        return $message->getFirstFrame()->getOpcode() === Frame::OP_PING;
+    }
+
+    public function process(Message $message, MessageProcessor $messageProcessor, ConnectionInterface $socket)
+    {
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createPongFrame($message->getContent()), $socket);
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createCloseFrame(), $socket);
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: A pong response should not close the connection !

File: src/Rfc6455/Message.php [hunk 8]
```diff
@@ -133,4 +135,21 @@ public function isOperation()
...
+    public function checkControlFrameSize()
+    {
+        $frame = $this->getFirstFrame();
+
+        if (!$frame->isControlFrame()) {
+            throw new ControlFrameException('It is not a control frame');
+        }
+
+        if ($frame->getPayloadLength() > Frame::MAX_CONTROL_FRAME_SIZE) {
+            throw new TooBigFrameException('A control frame cannot be larger than 125 bytes.');
```
Comment (Nek-) [inline src/Rfc6455/Message.php]: This exception is legit but should be throw by the frame itself. (also it should probably be a subclass of `TooBigFrameException` has its more a `TooBigControlFrameException`)

File: src/Rfc6455/Message.php [hunk 9]
```diff
@@ -133,4 +135,21 @@ public function isOperation()
...
+
+    /**
+     * @throws ControlFrameException
+     * @throws TooBigFrameException
+     */
+    public function checkControlFrameSize()
+    {
+        $frame = $this->getFirstFrame();
+
+        if (!$frame->isControlFrame()) {
```
Comment (Nek-) [inline src/Rfc6455/Message.php]: The most important check is the fact that a control frame [cannot be fragmented](https://tools.ietf.org/html/rfc6455#section-5.5). This mean that `count($this->frames) === 1`.

File: src/Rfc6455/Message.php (hunk 8 above)
Comment (Nek-) [inline src/Rfc6455/Message.php]: Here is a (fake) test case of what I think:

```php
public function testItThrowExceptionOnWrongControlFrame()
{
    $wrongFrame = // code to generate wrong frame
    $this->expectException('TooBigControlFrameException');
    $frame = new Frame($wrongFrame);
}
```

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    public function process(Message $message, MessageProcessor $messageProcessor, ConnectionInterface $socket)
+    {
+        try {
+            $message->checkControlFrameSize();
+        } catch (TooBigFrameException $e) {
+            $messageProcessor->timeout($socket);
+        }
+
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createPongFrame($message->getContent()), $socket);
+        $messageProcessor->write($messageProcessor->getFrameFactory()->createCloseFrame(), $socket);
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: NEIN NEIN NEIN NEIN NEIN NEIN NEIN

A pong does not imply a close frame.

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        return $message->getFirstFrame()->getOpcode() === Frame::OP_PING;
+
+    }
+
+    public function process(Message $message, MessageProcessor $messageProcessor, ConnectionInterface $socket)
+    {
+        try {
+            $message->checkControlFrameSize();
+        } catch (TooBigFrameException $e) {
+            $messageProcessor->timeout($socket);
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: wtf ?

The handler just handle *good* frame. If a frame is *wrong*, then the Frame cannot be initialized and throw an exception signaling that fact. As a consequence, the `Connection` close with a close message.

File: src/Rfc6455/Message.php (hunk 9 above)
Comment (Nek-) [inline src/Rfc6455/Message.php]: But IMO this should be throw by the Frame itself as the first frame knows if it's the last or not (thanks to the [first bit of the frame](https://tools.ietf.org/html/rfc6455#section-5.2))

File: tests/Woketo/Rfc6455/MessageHandler/PingFrameHandlerTest.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Test\Woketo\Rfc6455\MessageHandler;
+
+/**
+ * This file is a part of Woketo package.
+ *
+ * (c) Ci-tron <dev@ci-tron.org>
```
Comment (Nek-) [inline tests/Woketo/Rfc6455/MessageHandler/PingFrameHandlerTest.php]: `Nekland <dev@nekland.fr>` please :-°

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+use Nekland\Woketo\Rfc6455\Message;
+use Nekland\Woketo\Rfc6455\MessageProcessor;
+use React\Socket\ConnectionInterface;
+
+class PingFrameHandler implements Rfc6455MessageHandlerInterface
+{
+    public function supports(Message $message)
+    {
+        return $message->getFirstFrame()->getOpcode() === Frame::OP_PING;
+
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: useless line

File: src/Rfc6455/MessageHandler/PingFrameHandler.php [hunk 14]
```diff
@@ -0,0 +1,38 @@
+<?php
+/**
+ * This file is a part of Woketo package.
+ *
+ * (c) Ci-tron <dev@ci-tron.org>
```
Comment (Nek-) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: `Nekland <dev@nekland.fr>` please :-°

File: src/Rfc6455/Frame.php [hunk 15]
```diff
@@ -43,6 +43,8 @@ class Frame
     const CLOSE_UNEXPECTING_CONDITION   = 1011;
     // 1015 is reserved
 
+    const MAX_CONTROL_FRAME_SIZE = 125;
```
Comment (Nek-) [inline src/Rfc6455/Frame.php]: Nice, but add a link to `https://tools.ietf.org/html/rfc6455#section-5.5`

File: src/Exception/Frame/ControlFrameException.php [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * (c) Ci-tron <dev@ci-tron.org>
+ *
+ * For the full license, take a look to the LICENSE file
+ * on the root directory of this project
+ */
+
+namespace Nekland\Woketo\Exception\Frame;
+
+class ControlFrameException extends \Exception
```
Comment (Nek-) [inline src/Exception/Frame/ControlFrameException.php]: This should not inherit from `\Exception`.

File: src/Rfc6455/MessageHandler/PingFrameHandler.php (hunk 14 above)
Comment (folliked) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: OMG I copied a class of woketo and there are  throughout the project... You should also look at the configuration of your PhpStorm XD. I correct the ci-tron in my commit?

File: src/Rfc6455/Frame.php (hunk 15 above)
Comment (folliked) [inline src/Rfc6455/Frame.php]: ok

File: src/Rfc6455/MessageHandler/PingFrameHandler.php (hunk 14 above)
Comment (folliked) [inline src/Rfc6455/MessageHandler/PingFrameHandler.php]: https://github.com/Nekland/Woketo/search?utf8=%E2%9C%93&q=ci-tron

File: src/Rfc6455/Frame.php [hunk 17]
```diff
@@ -433,4 +452,20 @@ public function isValid() : bool
...
+     * @return bool
+     */
+    public function isControlFrame()
+    {
+        return (in_array($this->getOpcode(), [
+            self::OP_PONG,
+            self::OP_CLOSE,
+            self::OP_PING,
+        ])
+        || $this->getOpcode() >= 11 && $this->getOpcode() <= 15);
```
Comment (Nek-) [inline src/Rfc6455/Frame.php]: This condition is wrong. The Frame is a control frame when the opcode is `>= 8` (which is OP_CLOSE). (and it cannot be more than 15 by nature)

*Also careful with indentation, if this code is wrong, the indentation is also not good and parenthesis are useless.*

**Pull request merged**